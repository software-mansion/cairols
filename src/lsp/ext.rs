//! CairoLS extensions to the Language Server Protocol.

use std::path::PathBuf;

use lsp_types::notification::Notification;
use lsp_types::request::Request;
use lsp_types::{TextDocumentPositionParams, Url};
use serde::{Deserialize, Serialize};

/// Provides content of virtual file from the database.
pub(crate) struct ProvideVirtualFile;

#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
pub(crate) struct ProvideVirtualFileRequest {
    pub uri: Url,
}

#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
pub(crate) struct ProvideVirtualFileResponse {
    pub content: Option<String>,
}

impl Request for ProvideVirtualFile {
    type Params = ProvideVirtualFileRequest;
    type Result = ProvideVirtualFileResponse;
    const METHOD: &'static str = "vfs/provide";
}

/// Collects information about all Cairo crates that are currently being analyzed.
pub struct ViewAnalyzedCrates;

impl Request for ViewAnalyzedCrates {
    type Params = ();
    type Result = String;
    const METHOD: &'static str = "cairo/viewAnalyzedCrates";
}

/// Provides string with code after macros expansion.
pub struct ExpandMacro;

impl Request for ExpandMacro {
    type Params = TextDocumentPositionParams;
    type Result = Option<String>;
    const METHOD: &'static str = "cairo/expandMacro";
}

/// Notifies about corelib version mismatch.
#[derive(Debug)]
pub struct CorelibVersionMismatch;

impl Notification for CorelibVersionMismatch {
    type Params = String;
    const METHOD: &'static str = "cairo/corelib-version-mismatch";
}

/// Collects versions of LS and it's dependencies.
#[derive(Debug)]
pub struct ToolchainInfo;

#[derive(Serialize, Deserialize)]
pub struct PathAndVersion {
    pub path: PathBuf,
    pub version: String,
}

#[derive(Serialize, Deserialize)]
pub struct ToolchainInfoResponse {
    pub ls: PathAndVersion,
    pub scarb: Option<PathAndVersion>,
}

impl Request for ToolchainInfo {
    type Params = ();
    type Result = ToolchainInfoResponse;
    const METHOD: &'static str = "cairo/toolchainInfo";
}

#[cfg(feature = "testing")]
pub mod testing {
    use lsp_types::notification::Notification;

    /// Notifies about the end of project updating.
    #[derive(Debug)]
    pub struct ProjectUpdatingFinished;

    impl Notification for ProjectUpdatingFinished {
        type Params = ();
        const METHOD: &'static str = "cairo/projectUpdatingFinished";
    }

    /// Notifies about diagnostics generation which is beginning to calculate.
    #[derive(Debug)]
    pub struct DiagnosticsCalculationStart;

    impl Notification for DiagnosticsCalculationStart {
        type Params = ();
        const METHOD: &'static str = "cairo/diagnosticsCalculationStart";
    }

    /// Notifies about diagnostics generation which ended calculating.
    #[derive(Debug)]
    pub struct DiagnosticsCalculationFinish;

    impl Notification for DiagnosticsCalculationFinish {
        type Params = ();
        const METHOD: &'static str = "cairo/diagnosticsCalculationFinish";
    }
}

#[derive(Debug)]
pub struct ScarbPathMissing {}

impl Notification for ScarbPathMissing {
    type Params = ();
    const METHOD: &'static str = "scarb/could-not-find-scarb-executable";
}

#[derive(Debug)]
pub struct ScarbResolvingStart {}

impl Notification for ScarbResolvingStart {
    type Params = ();
    const METHOD: &'static str = "scarb/resolving-start";
}

#[derive(Debug)]
pub struct ScarbResolvingFinish {}

impl Notification for ScarbResolvingFinish {
    type Params = ();
    const METHOD: &'static str = "scarb/resolving-finish";
}
