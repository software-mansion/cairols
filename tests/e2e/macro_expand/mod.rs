use cairo_language_server::lsp::ext::ExpandMacro;
use itertools::Itertools;
use lsp_types::{TextDocumentIdentifier, TextDocumentPositionParams};
use serde::Serialize;

use crate::support::cairo_project_toml::CAIRO_PROJECT_TOML_2024_07;
use crate::support::cursor::peek_caret;
use crate::support::data::data;
use crate::support::quick_hash::quick_hash;
use crate::support::{cursors, sandbox};

#[test]
fn test_macro_expand_attribute() {
    check(data!("attribute.cairo.template"), "attribute")
}

#[test]
fn test_macro_expand_simple_inline() {
    check(data!("simple_inline.cairo.template"), "simple_inline")
}

#[test]
fn test_macro_expand_empty() {
    check(data!("empty.cairo.template"), "empty")
}

#[test]
fn test_macro_expand_derive() {
    check(data!("derive.cairo.template"), "derive")
}

fn check(code: String, case_name: &str) {
    let (cairo, cursors) = cursors(&code);

    let mut ls = sandbox! {
        files {
            "cairo_project.toml" => CAIRO_PROJECT_TOML_2024_07,
            "src/lib.cairo" => cairo.clone(),
        }
    };

    ls.open_all_cairo_files_and_wait_for_project_update();

    for (expansion, source_contexts) in cursors
        .carets()
        .into_iter()
        .map(|position| {
            let source_context = peek_caret(&cairo, position);

            let macro_expansion = ls.send_request::<ExpandMacro>(TextDocumentPositionParams {
                position,
                text_document: TextDocumentIdentifier { uri: ls.doc_id("src/lib.cairo").uri },
            });

            let expansion =
                macro_expansion.unwrap_or_else(|| String::from("No expansion information.\n"));

            (expansion, source_context)
        })
        .into_group_map()
    {
        let hash = quick_hash(&expansion);
        let source_contexts = source_contexts.join("\n");
        let report = MacroExpandReport { source_contexts, expansion };

        insta::assert_toml_snapshot!(format!("{case_name}#{hash:x}"), report);
    }
}

#[derive(Serialize)]
struct MacroExpandReport {
    source_contexts: String,
    expansion: String,
}
