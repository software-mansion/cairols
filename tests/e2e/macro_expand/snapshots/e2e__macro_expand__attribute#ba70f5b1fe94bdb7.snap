---
source: tests/e2e/macro_expand/mod.rs
expression: report
---
source_contexts = '''
        print<caret>ln!("a");
'''
expansion = '''
// lib.cairo
// ---------

{
    let mut __formatter_for_print_macros__: core::fmt::Formatter = core::traits::Default::default();
    core::result::ResultTrait::<
        (), core::fmt::Error,
    >::unwrap(
        {
            core::byte_array::ByteArrayTrait::append_word(
                ref __formatter_for_print_macros__.buffer, 0x610a, 2,
            );
            core::result::Result::<(), core::fmt::Error>::Ok(())
        },
    );
    core::debug::print_byte_array_as_string(@__formatter_for_print_macros__.buffer);
}'''
